### PR TITLE
perf: Remove some redundant copying of batches

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

We use `CopyExec` to wrap inputs to operators that can cache batches. This is to avoid corruption issues if the input batches come from a `ScanExec` where arrays are re-used, or from some other operators that may be wrapping `ScanExec` and that can pass through input batches unchanged (projection, limit, or filter).

`CopyExec` also unpacks dictionary arrays (not just primitives, but Utf8 as well).

When we have nested joins, I see that `CopyExec` is processing the output of the nested joins and this is redundant because joins do not re-use arrays.

This PR aims to avoid deep copies of arrays in some cases where it is not necessary.

## How are these changes tested?

Existing tests

### CODE REVIEW

File: native/core/src/execution/operators/copy.rs [hunk 1]
```diff
@@ -200,3 +217,56 @@ impl RecordBatchStream for CopyStream {
         self.schema.clone()
     }
 }
+
+/// Copy an Arrow Array
+fn copy_array(array: &dyn Array) -> ArrayRef {
```
Comment (andygrove) [inline native/core/src/execution/operators/copy.rs:222]: this function was moved from `mod.rs` with no changes

File: native/core/src/execution/operators/copy.rs [hunk 2]
```diff
@@ -200,3 +217,56 @@ impl RecordBatchStream for CopyStream {
...
+        make_array(mutable.freeze())
+    }
+}
+
+/// Copy an Arrow Array or cast to primitive type if it is a dictionary array.
+/// This is used for `CopyExec` to copy/cast the input array. If the input array
+/// is a dictionary array, we will cast the dictionary array to primitive type
+/// (i.e., unpack the dictionary array) and copy the primitive array. If the input
+/// array is a primitive array, we simply copy the array.
+fn copy_or_unpack_array(array: &Arc<dyn Array>, mode: &CopyMode) -> Result<ArrayRef, ArrowError> {
```
Comment (andygrove) [inline native/core/src/execution/operators/copy.rs:258]: this function was moved from `mod.rs` but was also modified

File: native/core/src/execution/datafusion/planner.rs [hunk 3]
```diff
@@ -1205,15 +1210,15 @@ impl PhysicalPlanner {
         // to copy the input batch to avoid the data corruption from reusing the input
         // batch.
         let left = if can_reuse_input_batch(&left) {
-            Arc::new(CopyExec::new(left))
+            Arc::new(CopyExec::new(left, CopyMode::UnpackOrDeepCopy))
```
Comment (andygrove) [inline native/core/src/execution/datafusion/planner.rs:1213]: Note that `can_reuse_input_batch` returns `true` in fewer cases now, so we are less likely to create the deep-copy version of CopyExec.

Previously we would create a `CopyExec` around any projection or filter, but now we only do so if the projection or filter is wrapping a scan.

File: native/core/src/execution/datafusion/planner.rs [hunk 4]
```diff
@@ -1205,15 +1210,15 @@ impl PhysicalPlanner {
...
         } else {
-            left
+            Arc::new(CopyExec::new(left, CopyMode::UnpackOrClone))
```
Comment (andygrove) [inline native/core/src/execution/datafusion/planner.rs:1215]: We need to unpack dictionaries before joins currently

**Comment by andygrove** [comment]
I did see a small performance improvement with this change when running TPC-DS but not enough to be statistically significant.

File: native/core/src/execution/datafusion/planner.rs [hunk 5]
```diff
@@ -1775,10 +1780,14 @@ impl From<ExpressionError> for DataFusionError {
 /// modification. This is used to determine if we need to copy the input batch to avoid
 /// data corruption from reusing the input batch.
 fn can_reuse_input_batch(op: &Arc<dyn ExecutionPlan>) -> bool {
-    op.as_any().is::<ScanExec>()
```
Comment (viirya) [inline native/core/src/execution/datafusion/planner.rs:1778]: Hmm, but our native scan actually reuses batches.

Comment (andygrove) [reply native/core/src/execution/datafusion/planner.rs:1778]: yes, it is still being checked here:

```rust
 } else {
         op.as_any().is::<ScanExec>()
     }
```

Comment (viirya) [reply native/core/src/execution/datafusion/planner.rs:1778]: Ah, okay, I missed it.

**Review by viirya — APPROVED** [review]
Makes sense to me. Thanks @andygrove

**Comment by andygrove** [comment]
Thanks for the review @viirya

**Pull request merged**